File: python/LengthOfLastWord.py
# Link: https://leetcode.com/problems/length-of-last-word/
class Solution:
    def lengthOfLastWord(self, s: str) -> int:
        last_word_size = 0
        string_size = len(s)
        i = 0

        while i<string_size:
            if s[i] != " ":
                last_word_size+=1
                i+=1

            else:
                while (i<string_size and s[i] == " "): i+=1

                if i == string_size:
                    return last_word_size
                
                else:
                    last_word_size = 0
        
        return last_word_size 

'''
    Complexity analysis:
    * Time: O(n)
        Since we loop through all the characters in the string, once, the time complexity grows linearly
        with the size of the input.
    * Space: O(1)
        We only use a constant number of variables, so the space complexity is constant.
'''

File: javascript/LengthOfLastWord.js
// Leetcode: https://leetcode.com/problems/length-of-last-word/
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLastWord = function(s) {
    let lastWordSize = 0;
    let stringSize = s.length;
    let i = 0;

    while (i<stringSize) {
        if (s[i] != " ") {
            lastWordSize++;
            i++;
        }
        else {
            while (i<stringSize && s[i]==" ") i++;

            if (i==stringSize) {
                return lastWordSize;
            } else {
                lastWordSize = 0;
            }
        }
    }
    return lastWordSize;
};
/*
 * Complexity analysis
 * Time: O(n)
 * Since we iterate through all the string once and once only
 * we have a linear dependency on the size of the string.
 * Space: O(1)
 * We use a constant number of variables to keep track of the size of the last word.
*/
